Dmf/Modules.Library/Dmf_Time.c

```diff
@@ -0,0 +1,426 @@
+#include "Dmf_Time.tmh"
+#endif
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+// Module Private Enumerations and Structures
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+//
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+// Module Private Context
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+//
+
+typedef struct _DMF_CONTEXT_Time
+{
+    // Get Performance Frequency during module Open. This value will not change. 
+    //
+    LARGE_INTEGER PerformanceFrequency;
+} DMF_CONTEXT_Time;
+
+// This macro declares the following function:
+// DMF_CONTEXT_GET()
+//
+DMF_MODULE_DECLARE_CONTEXT(Time)
+
+// This macro declares the following function:
+// DMF_CONFIG_GET()
+//
+DMF_MODULE_DECLARE_NO_CONFIG(Time)
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+// DMF Module Support Code
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+//
+
+#define NANOSECONDS_TO_MILLISECONDS         (1000000U)
+#define SECONDS_TO_NANOSECONDS              (1000000000U)
+
+__forceinline
+NTSTATUS
+ElapsedTimeInNanosecondsGet(
+    _In_ LONGLONG CurrentTick,
+    _In_ LONGLONG LastQueryTick,
+    _In_ LONGLONG PerformanceFrequency,
+    _Out_ LONGLONG* ElapsedTimeInNanoseconds
+    )
+/*++
+
+Routine Description:
+
+    Calculate elapsed time in nanoseconds given the current and last tick count.
+
+Arguments:
+
+    CurrentTick - Current tick count.
+    LastQueryTick - Last tick count.
+    PerformanceFrequency - performance-counter frequency.
+    ElapsedTimeInNanoseconds - Return the elapsed time in nanoseconds.
+
+Return Value:
+
+    NTSTATUS
+
+--*/
+{
+    NTSTATUS ntStatus;
+    LONGLONG elaspedTicks;
+    LONGLONG result;
+
+    DmfAssert(PerformanceFrequency);
+
+    ntStatus = STATUS_UNSUCCESSFUL;
+
+    elaspedTicks = CurrentTick - LastQueryTick;
+
+#ifdef DMF_KERNEL_MODE
+    ntStatus = RtlLong64Mult(elaspedTicks,
+                             SECONDS_TO_NANOSECONDS, 
+                             &result);
+    if (!NT_SUCCESS(ntStatus))
+    {
+        DmfAssert(false);
+        TraceEvents(TRACE_LEVEL_ERROR, DMF_TRACE, "Integer overflow when performing multiplication of %lld and %lld", elaspedTicks, SECONDS_TO_NANOSECONDS);
+        goto Exit;
+    }
+#else
+    HRESULT hResult;
+
+    hResult = Long64Mult(elaspedTicks,
+                         SECONDS_TO_NANOSECONDS,
+                         &result);
+    if (hResult != S_OK)
+    {
+        DmfAssert(false);
+        TraceEvents(TRACE_LEVEL_ERROR, DMF_TRACE, "Integer overflow when performing multiplication of %lld and %lld", elaspedTicks, SECONDS_TO_NANOSECONDS);
+        ntStatus = STATUS_INTEGER_OVERFLOW;
+        goto Exit;
+    }
+#endif

```

Comment:
where are we making status success for UserMode?

Reply:
fixed